[GStreamer][MSE] Support text tracks in MSE

vivienne-w opened the pull request:
#### e1f55e8d9ee93718eb1013f6ded952a00887990f
<pre>
[GStreamer][MSE] Support text tracks in MSE
<a href="https://bugs.webkit.org/show_bug.cgi?id=281343">https://bugs.webkit.org/show_bug.cgi?id=281343</a>

Reviewed by Alicia Boya Garcia and Xabier Rodriguez-Calvar.

Add support for text tracks in MSE, in particular WebVTT tracks.

In AppendPipeline, we reencode any text tracks to WebVTT by plugging webvttenc if necessary.
(Optional encoders were implemented in: <a href="https://commits.webkit.org/290605@main">https://commits.webkit.org/290605@main</a> )

As WebKit takes care of subtitles anyway, these tracks don&apos;t flow through the playback pipeline
and are instead directly handled by InbandTextTrackPrivateGStreamer. We remove any text tracks
from the stream collection, and so webkitmediasrc will never create pads for them.

* LayoutTests/media/media-source/content/counting-video-vttsimple-manifest.json: Added.
* LayoutTests/media/media-source/content/counting-video-vttsimple.mp4: Added.
* LayoutTests/media/media-source/media-source-vttsimple-expected.txt: Added.
* LayoutTests/media/media-source/media-source-vttsimple.html: Added.
  Similar to media-managedmse-webvtt-track.html, but without using managed MSE.

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::initializeDecoders):
  Always support webvtt in MSE.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
  Allow text tracks for MSE.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::parseDemuxerSrcPadCaps):
  Assign StreamType::Text to application/x-subtitle-webvtt.
(WebCore::AppendPipeline::appsinkNewSample):
  Handle text samples in InbandTextTrackPrivateGStreamer.
(WebCore::AppendPipeline::didReceiveInitializationSegment):
  Add text track info to InitializationSegment.
(WebCore::createOptionalEncoderForFormat):
  Plug webvttenc, if necessary.
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat):
  Add missing underscore to element name.
(WebCore::AppendPipeline::Track::emplaceOptionalEncoderForFormat):
  Ditto.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::sourceSetup):
  Refactored code in common with start Source to emitStreams.
(WebCore::MediaPlayerPrivateGStreamerMSE::startSource):
  Ditto.
(WebCore::MediaPlayerPrivateGStreamerMSE::emitStreams):
  Filters and emits streams.
(WebCore::filterOutRepeatingTracks):
  Deleted. Filtering is done in emitStreams now.
(WebCore::MediaPlayerPrivateGStreamerMSE::sourceSetup):
  Don&apos;t include text tracks in the stream collection.
(WebCore::MediaPlayerPrivateGStreamerMSE::startSource):
  Ditto.

Canonical link: <a href="https://commits.webkit.org/291495@main">https://commits.webkit.org/291495@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f9dd46df9ef62736862919e4f30f598cdbda80d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12217 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/1840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97657 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43180 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94719 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20672 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70962 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28398 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9466 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83897 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51294 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9152 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1523 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42507 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79474 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99682 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19720 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14522 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79970 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19970 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79788 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79270 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19739 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23792 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1070 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12735 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19704 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24881 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19391 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22851 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21132 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -518,8 +518,12 @@ void GStreamerRegistryScanner::initializeDecoders(const GStreamerRegistryScanner
 
-    if (m_isMediaSource)
+    if (m_isMediaSource) {
+        // FIXME ditry hack to ensure WebVTT will always be supported in MSE, since we dont actually push any WebVTT to the playback pipeline
```
ditry -> dirty, don't  . Period at the end .

But anyway, I don't think about this as a hack. If it does not come into the playback pipeline, ensuring we answer that we support it is not a hack, IMHO.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -373,8 +373,9 @@ void AppendPipeline::appsinkNewSample(const Track& track, GRefPtr<GstSample>&& s
-    if (!GST_BUFFER_PTS_IS_VALID(buffer)) {
+    if (!GST_BUFFER_PTS_IS_VALID(buffer) && track.streamType != Text) {
         // When demuxing Vorbis, matroskademux creates several PTS-less frames with header information. We don't need those.
+        // However, for text tracks, thats likely our WebVTT header, which is required for WebVTTParser.
```
that's

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -530,6 +534,14 @@ void AppendPipeline::didReceiveInitializationSegment()
+            SourceBufferPrivateClient::InitializationSegment::TextTrackInformation info;
+            info.track = static_cast<InbandTextTrackPrivateGStreamer*>(track->webKitTrack.get());
+            info.description = GStreamerMediaDescription::create(track->finalCaps);
+            initializationSegment.textTracks.append(info);
```
WTFMove(info)

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -405,7 +421,8 @@ void MediaPlayerPrivateGStreamerMSE::sourceSetup(GstElement* sourceElement)
         m_tracks = filterOutRepeatingTracks(m_tracks);
-        webKitMediaSrcEmitStreams(WEBKIT_MEDIA_SRC(m_source.get()), m_tracks);
+        auto playbackTracks = filterOutTextTracks(m_tracks);
+        webKitMediaSrcEmitStreams(WEBKIT_MEDIA_SRC(m_source.get()), playbackTracks);
```
Seems like `webKitMediaSrcEmitStreams()` can now take a `Vector<>&&` (each track could be moved to the `Stream`)? If so, WTFMove can be used here.

philn replied:
`webKitMediaSrcEmitStreams()` would need to be updated as well.

ntrrgc replied:
> Seems like `webKitMediaSrcEmitStreams()` can now take a `Vector<>&&` (each track could be moved to the Stream)?

Although you can technically do this, it is a bad idea API-design-wise. Having a parameter of type `Vector<>&&` communicates that the callee will take ownership of the vector. From that it would be easily to wrongly assume that WebKitMediaSrc keeps a `Vector<MediaSourceTrackGStreamer>`, which it intentionally does not.

On the other hand, having a parameter of type `const Vector<>&` communicates that `webKitMediaSrcEmitStreams()` will borrow the vector, read the contents and not use it once the function has returned.

This actually matches what happens in `webKitMediaSrcEmitStreams()`: the vector of MediaSourceTrackGStreamer is used to initialize `source->priv->streams`, which is of a completely different type (`TrackIDHashMap<RefPtr<Stream>>`).

vivienne-w replied:
I agree with Alicia now, so I changed `webKitMediaSrcEmitStreams` back.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
```diff
@@ -518,8 +518,12 @@ void GStreamerRegistryScanner::initializeDecoders(const GStreamerRegistryScanner
 
-    if (m_isMediaSource)
+    if (m_isMediaSource) {
+        // This ensures WebVTT will always be supported in MSE, since we don't actually push any WebVTT to the playback pipeline.
```
I would add this to the comment too:
```
// WebVTT is decoded directly by WebKit.
```

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -307,16 +307,16 @@ std::tuple<GRefPtr<GstCaps>, AppendPipeline::StreamType, FloatSize> AppendPipeli
     auto originalMediaType = capsMediaType(demuxerSrcPadCaps);
     auto& gstRegistryScanner = GStreamerRegistryScannerMSE::singleton();
-    if (!gstRegistryScanner.isCodecSupported(GStreamerRegistryScanner::Configuration::Decoding, originalMediaType.toStringWithoutCopying())) {
+    if (doCapsHaveType(demuxerSrcPadCaps, GST_TEXT_CAPS_TYPE_PREFIX) || originalMediaType == "application/x-subtitle-vtt"_s) {
```
Unless I'm missing something this wouldn't work with if the muxed subtitle format is something other than WebVTT, as we're checking the caps of the demuxer srcpad which is upstream of webvttenc.

vivienne-w replied:
You're right, I forgot about other formats here. However, since `webvttenc` expects `text/x-raw`, and we replace it with an `identity` when we get WebVTT already, this does actually work for those cases.

For anything else, I'd actually need to add `subparse` to the AppendPipeline.

ntrrgc replied:
Does it? Because you're working with `demuxerSrcPad`, but isn't that before webvttenc?

vivienne-w replied:
Yes, but where is the problem with that? That pad can either have `text/x-raw` or WebVTT, and in both cases we consider it a text track.

vivienne-w replied:
Either way, I see now that supporting more formats is a bit more involved than I realized, so I'd prefer to split that out into a follow-up patch, and maybe pick some specific formats to add support for. (For now, at least)

Does that work for you?

ntrrgc replied:
> Yes, but where is the problem with that? That pad can either have text/x-raw or WebVTT, and in both cases we consider it a text track.

Oh, it's an `||`. For some reason I read it as a `&&`. It's fine then, sorry for the confusion.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -409,7 +410,11 @@ void AppendPipeline::appsinkNewSample(const Track& track, GRefPtr<GstSample>&& s
+        const auto textTrack = static_cast<InbandTextTrackPrivateGStreamer*>(track.webKitTrack.get());
+        textTrack->handleSample(GRefPtr(mediaSample->sample()));
+    } else
+        m_sourceBufferPrivate.didReceiveSample(mediaSample.get());
```
I'm surprised it works like this.

`didReceiveSample()` triggers the processing in `SourceBufferPrivate::processMediaSample()` which is generic for all sample types. `sample->offsetTimestampsBy()` is called there, for instance. Buffered ranges are also updated there.

I think we should be calling `didReceiveSample()` for text samples as well unless we have a good reason not to. The macOS port does this as well. Try to remove the `else` so that both `handleSample()` and `didReceiveSample()` are used and check if anything breaks.

ntrrgc replied:
Just realizing now, you may want to not call `didReceiveSample` for frames without timestamp.

vivienne-w replied:
It used to cause a crash, but after trying it again its working now

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -776,6 +789,9 @@ GRefPtr<GstElement> createOptionalEncoderForFormat([[maybe_unused]] GstBin* bin,
 
     const char* elementClass = "identity";
 
+    // FIXME scenarios we don't handle yet:
```
```
// FIXME scenarios we haven't tested yet:
```

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -405,7 +421,8 @@ void MediaPlayerPrivateGStreamerMSE::sourceSetup(GstElement* sourceElement)
         m_tracks = filterOutRepeatingTracks(m_tracks);
-        webKitMediaSrcEmitStreams(WEBKIT_MEDIA_SRC(m_source.get()), m_tracks);
+        auto playbackTracks = filterOutTextTracks(m_tracks);
+        webKitMediaSrcEmitStreams(WEBKIT_MEDIA_SRC(m_source.get()), WTFMove(playbackTracks));
```
I don't like the code duplication here, and we should be able to get rid of it, but that will be a different patch.

vivienne-w replied:
Since `sourceSetup` and `startSource` are running the exact same 3 lines here, I could refactor those out into a new method, and replace both the filter functions too.

ntrrgc replied:
I looked into this the other day, and I think the filtering would only need to be done in one place.

This is the diagram I made while analyzing it:

![startPlaybackIfHasAllTracks diagram](https://github.com/user-attachments/assets/ac3748ae-b4c6-453c-9c18-a04e0ffe0e00)

ntrrgc reviewed (changes requested):
It looks promising.

We should add a test for a MP4 file that doesn't use WebVTT to make sure that works in the GStreamer port as well, and that it keeps working in the future.

Please add `[MSE]` in the title of the commit.

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -409,6 +410,11 @@ void AppendPipeline::appsinkNewSample(const Track& track, GRefPtr<GstSample>&& s
+        textTrack->handleSample(GRefPtr(mediaSample->sample()));
+    }
+
     m_sourceBufferPrivate.didReceiveSample(mediaSample.get());
```
We must not call `didReceiveSample()` if we don't have valid timestamps, I checked:

If we call `didReceiveSample()` with a sample that has no timestamps, it will end up in `SourceBufferPrivate::processMediaSample()`, and inside there `sample->presentationTime()` is called and arithmetic is done on it. That's going to lead to problems if the presentationTime is set as invalid.

For video tracks we avoid this problem with the early return from before:

```
   if (!GST_BUFFER_PTS_IS_VALID(buffer) && track.streamType != Text) {
       // [...]
       return;
```

However, since we don't do the early return for text tracks in order to handle the WebVTT headers, we need to guard against invalid PTS again here.

ntrrgc reviewed (approved):
r+ with one change.

I still want a LayoutTest doing the same test but with a text track format other than WebVTT so that support for that is properly tested. However I'm okay with landing this current version and having that on a different patch too, especially if that effort uncovers any otherwise unrelated bugs, as often happens.

vivienne-w committed:
[GStreamer][MSE] Support text tracks in MSE
https://bugs.webkit.org/show_bug.cgi?id=281343

Reviewed by Alicia Boya Garcia and Xabier Rodriguez-Calvar.

Add support for text tracks in MSE, in particular WebVTT tracks.

In AppendPipeline, we reencode any text tracks to WebVTT by plugging webvttenc if necessary.
(Optional encoders were implemented in: https://commits.webkit.org/290605@main )

As WebKit takes care of subtitles anyway, these tracks don't flow through the playback pipeline
and are instead directly handled by InbandTextTrackPrivateGStreamer. We remove any text tracks
from the stream collection, and so webkitmediasrc will never create pads for them.

* LayoutTests/media/media-source/content/counting-video-vttsimple-manifest.json: Added.
* LayoutTests/media/media-source/content/counting-video-vttsimple.mp4: Added.
* LayoutTests/media/media-source/media-source-vttsimple-expected.txt: Added.
* LayoutTests/media/media-source/media-source-vttsimple.html: Added.
  Similar to media-managedmse-webvtt-track.html, but without using managed MSE.

* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::initializeDecoders):
  Always support webvtt in MSE.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTracks):
  Allow text tracks for MSE.
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::parseDemuxerSrcPadCaps):
  Assign StreamType::Text to application/x-subtitle-webvtt.
(WebCore::AppendPipeline::appsinkNewSample):
  Handle text samples in InbandTextTrackPrivateGStreamer.
(WebCore::AppendPipeline::didReceiveInitializationSegment):
  Add text track info to InitializationSegment.
(WebCore::createOptionalEncoderForFormat):
  Plug webvttenc, if necessary.
(WebCore::AppendPipeline::Track::emplaceOptionalParserForFormat):
  Add missing underscore to element name.
(WebCore::AppendPipeline::Track::emplaceOptionalEncoderForFormat):
  Ditto.
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::sourceSetup):
  Refactored code in common with start Source to emitStreams.
(WebCore::MediaPlayerPrivateGStreamerMSE::startSource):
  Ditto.
(WebCore::MediaPlayerPrivateGStreamerMSE::emitStreams):
  Filters and emits streams.
(WebCore::filterOutRepeatingTracks):
  Deleted. Filtering is done in emitStreams now.
(WebCore::MediaPlayerPrivateGStreamerMSE::sourceSetup):
  Don't include text tracks in the stream collection.
(WebCore::MediaPlayerPrivateGStreamerMSE::startSource):
  Ditto.

Canonical link: https://commits.webkit.org/291495@main
